### PR TITLE
Using constraints with two index sets for dealii-9.6 only.

### PR DESCRIPTION
**Pull request by vyushut**
Guarding #if DEAL_II_VERSION_GTE(9,6,0) blocks were added to all "constraints.reinit"

### CODE REVIEW

File: source/mesh_deformation/interface.cc [hunk 1]
```diff
@@ -1385,8 +1400,11 @@ namespace aspect
       // after setup_dofs(), as is done, for instance, during mesh
       // refinement.
       mesh_vertex_constraints.clear();
-      mesh_vertex_constraints.reinit(mesh_locally_relevant);
-
+#if DEAL_II_VERSION_GTE(9,6,0)
+        mesh_vertex_constraints.reinit(dof_handler.locally_owned_dofs(), mesh_locally_relevant);
```
Comment (tjhei) [inline source/mesh_deformation/interface.cc:1404]: This needs to be 
```suggestion
        mesh_vertex_constraints.reinit(mesh_deformation_dof_handler.locally_owned_dofs(), mesh_locally_relevant);
```

File: source/mesh_deformation/interface.cc [hunk 2]
```diff
@@ -685,8 +689,11 @@ namespace aspect
       // because this object is used for updating the displacement in
       // compute_mesh_displacements().
       mesh_velocity_constraints.clear();
-      mesh_velocity_constraints.reinit(mesh_locally_relevant);
-
+#if DEAL_II_VERSION_GTE(9,6,0)
+        mesh_velocity_constraints.reinit(dof_handler.locally_owned_dofs(), mesh_locally_relevant);
```
Comment (tjhei) [inline source/mesh_deformation/interface.cc:693]:
```suggestion
        mesh_velocity_constraints.reinit(mesh_deformation_dof_handler.locally_owned_dofs(), mesh_locally_relevant);
```

File: source/mesh_deformation/interface.cc [hunk 3]
```diff
@@ -586,7 +586,11 @@ namespace aspect
 
       // Now construct the mesh displacement constraints
       mesh_velocity_constraints.clear();
-      mesh_velocity_constraints.reinit(mesh_locally_relevant);
+#if DEAL_II_VERSION_GTE(9,6,0)
+        mesh_velocity_constraints.reinit(dof_handler.locally_owned_dofs(), mesh_locally_relevant);
```
Comment (tjhei) [inline source/mesh_deformation/interface.cc:590]:
```suggestion
        mesh_velocity_constraints.reinit(mesh_deformation_dof_handler.locally_owned_dofs(), mesh_locally_relevant);
```

File: source/simulator/core.cc [hunk 4]
```diff
@@ -1449,7 +1461,12 @@ namespace aspect
 
     // Reconstruct the constraint-matrix:
     constraints.clear();
-    constraints.reinit(introspection.index_sets.system_relevant_set);
+#if DEAL_II_VERSION_GTE(9,6,0)
+      constraints.reinit (dof_handler.locally_owned_dofs(),
+                                    introspection.index_sets.system_relevant_set);
```
Comment (tjhei) [inline source/simulator/core.cc:1466]: did you run ``make indent`` and have astyle installed?

File: source/simulator/stokes_matrix_free.cc [hunk 5]
```diff
@@ -2476,8 +2476,11 @@ namespace aspect
             if (!no_flux_boundary.empty())
               {
                 AffineConstraints<double> user_level_constraints;
+#if DEAL_II_VERSION_GTE(9,6,0)
+                user_level_constraints.reinit(dof_handler.locally_owned_dofs(), relevant_dofs);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:2480]:
```suggestion
                user_level_constraints.reinit(dof_handler_v.locally_owned_dofs(), relevant_dofs);
```

**Review by tjhei — CHANGES_REQUESTED** [review]

**Comment by tjhei** [comment]
alternative to #5395

File: source/simulator/core.cc [hunk 6]
```diff
@@ -657,7 +657,12 @@ namespace aspect
     // if the set of constraints has changed. If it did, we need to update the sparsity patterns.
     AffineConstraints<double> new_current_constraints;
     new_current_constraints.clear ();
-    new_current_constraints.reinit (introspection.index_sets.system_relevant_set);
+#if DEAL_II_VERSION_GTE(9,6,0)
+      new_current_constraints.reinit (dof_handler.locally_owned_dofs(),
+                                    introspection.index_sets.system_relevant_set);
```
Comment (tjhei) [inline source/simulator/core.cc:662]: indentation looks wrong.

**Review by tjhei — APPROVED** [review]
Tester fails because the required deal.II fix has not been merged. I think this PR is good to go though.

**Comment by gassmoeller** [comment]
Thanks for working on this fix @vyushut!

I tried this branch with the latest deal.II development, but I still see some test failures (they should show in the tester as well once it is done running). In particular the test `particle_periodic_boundaries_dc` fails (with MPI:2, it works in serial runs), which leads me to believe something is still broken with our constraints for discontinuous fields. `particle_periodic_boundaries` which is an exact copy, but without the discontinuous compositional fields, passes. @tjhei do you know of any fixes for deal.II that are still left to do?

Backtrace of the thrown unknown exception:

```
-----------------------------------------------------------------------------
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 2.6.0-pre (two-index_constraints, 3fedd2b83)
--     . using deal.II 9.6.0-pre (master, edc7d5c3ce)
--     .       with 64 bit indices and vectorization level 1 (128 bits)
--     . using Trilinos 13.2.0
--     . using p4est 2.3.2
--     . running in DEBUG mode
--     . running with 2 MPI processes
-----------------------------------------------------------------------------

-----------------------------------------------------------------------------
-- For information on how to cite ASPECT, see:
--   https://aspect.geodynamics.org/citing.html?ver=2.6.0-pre&dg=1&particles=1&sha=3fedd2b83&src=code
-----------------------------------------------------------------------------
Number of active cells: 64 (on 4 levels)
Number of degrees of freedom: 1,524 (578+81+289+576)

*** Timestep 0:  t=0 seconds, dt=0 seconds
   Skipping temperature solve because RHS is zero.
   Rebuilding Stokes preconditioner...
Thread 1 "aspect" hit Catchpoint 1 (exception thrown), 0x00007fffdb6ae4a1 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007fffdb6ae4a1 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00007fffdb6a2617 in __cxa_bad_cast () from /lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007fffdbc0b58b in Epetra_CrsMatrix::CheckSizes (this=<optimized out>, Source=...)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/epetra/src/Epetra_CrsMatrix.cpp:2153
#3  0x00007fffdbc84736 in Epetra_DistObject::DoTransfer (this=0x55555bdf6690, A=..., CombineMode=Insert, NumSameIDs=306, NumPermuteIDs=0, 
    NumRemoteIDs=272, NumExportIDs=0, PermuteToLIDs=0x0, PermuteFromLIDs=0x0, RemoteLIDs=0x55555ba34140, ExportLIDs=0x0, LenExports=@0x55555bdf66d0: 0, 
    Exports=@0x55555bdf66c0: 0x0, LenImports=@0x55555bdf66d4: 0, Imports=@0x55555bdf66c8: 0x0, Distor=..., DoReverse=false, Indexor=0x0)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/epetra/src/Epetra_DistObject.cpp:245
#4  0x00007fffdbc854d4 in Epetra_DistObject::Import (this=<optimized out>, A=..., Importer=..., CombineMode=CombineMode@entry=Insert, 
    Indexor=Indexor@entry=0x0)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/epetra/src/Epetra_DistObject.cpp:134
#5  0x00007fffdb4e4f77 in Amesos_Klu::ExportToSerial (this=0x55555bcd9a70)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/teuchos/core/src/Teuchos_RCP.hpp:379
#6  0x00007fffdb4eb492 in Amesos_Klu::SymbolicFactorization (this=0x55555bcd9a70)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/amesos/src/Amesos_Klu.cpp:664
#7  0x00007fffdb4ea862 in Amesos_Klu::NumericFactorization (this=0x55555bcd9a70)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/amesos/src/Amesos_Klu.cpp:696
#8  0x00007fffdadb13f4 in ML_Amesos_Gen (ml=ml@entry=0x55555bbd5318, curr_level=curr_level@entry=0, choice=choice@entry=0, MaxProcs=<optimized out>, 
    MaxProcs@entry=-1, AddToDiag=AddToDiag@entry=9.9999999999999998e-13, splitComm=splitComm@entry=1, Amesos_Handle=Amesos_Handle@entry=0x55555bdeeee0)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/ml/src/Utils/ml_amesos_wrap.cpp:256
#9  0x00007fffdadad154 in ML_Gen_Smoother_Amesos (ml=0x55555bbd5318, nl=0, AmesosSolver=0, MaxProcs=-1, AddToDiag=9.9999999999999998e-13, SplitComm=1)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/ml/src/Utils/ml_amesos.c:90
#10 0x00007fffdae35423 in ML_Epetra::MultiLevelPreconditioner::SetSmoothers (this=this@entry=0x55555bcc9cd0, 
    keepFineLevelSmoother=keepFineLevelSmoother@entry=false)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/ml/src/Utils/ml_MultiLevelPreconditioner_Smoothers.cpp:1869
#11 0x00007fffdadedb19 in ML_Epetra::MultiLevelPreconditioner::ComputePreconditioner (this=<optimized out>, CheckPreconditioner=<optimized out>)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/ml/src/Utils/ml_MultiLevelPreconditioner.cpp:2519
#12 0x00007fffdadf3eea in ML_Epetra::MultiLevelPreconditioner::MultiLevelPreconditioner (this=0x55555bcc9cd0, inRowMatrix=..., List=..., 
    ComputePrec=<optimized out>, __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /home/renegassmoeller/software/tmp/unpack/Trilinos-trilinos-release-13-2-0/packages/ml/src/Utils/ml_MultiLevelPreconditioner.cpp:356
#13 0x00007fffef9cd97b in dealii::TrilinosWrappers::PreconditionAMG::initialize (this=0x55555b47b0f0, matrix=..., ml_parameters=...)
    at /home/renegassmoeller/software/dealii/source/lac/trilinos_precondition_ml.cc:270
#14 0x00007fffef9cd734 in dealii::TrilinosWrappers::PreconditionAMG::initialize (this=0x55555b47b0f0, matrix=..., additional_data=...)
    at /home/renegassmoeller/software/dealii/source/lac/trilinos_precondition_ml.cc:241
#15 0x00007fffef9cd69f in dealii::TrilinosWrappers::PreconditionAMG::initialize (this=0x55555b47b0f0, matrix=..., additional_data=...)
    at /home/renegassmoeller/software/dealii/source/lac/trilinos_precondition_ml.cc:225
#16 0x00005555585cf3e7 in aspect::Simulator<2>::build_stokes_preconditioner (this=0x7fffffff8d60)
    at /home/renegassmoeller/software/aspect/source/simulator/assembly.cc:514
#17 0x0000555557ef8117 in aspect::Simulator<2>::assemble_and_solve_stokes (this=0x7fffffff8d60, initial_nonlinear_residual=@0x7fffffff8510: 0, 
    nonlinear_residual=0x0) at /home/renegassmoeller/software/aspect/source/simulator/solver_schemes.cc:397
#18 0x0000555557ef84fc in aspect::Simulator<2>::solve_single_advection_single_stokes (this=0x7fffffff8d60)
    at /home/renegassmoeller/software/aspect/source/simulator/solver_schemes.cc:767
#19 0x0000555557bb8a9f in aspect::Simulator<2>::solve_timestep (this=0x7fffffff8d60)
    at /home/renegassmoeller/software/aspect/source/simulator/core.cc:1852
#20 0x0000555557bb5911 in aspect::Simulator<2>::run (this=0x7fffffff8d60) at /home/renegassmoeller/software/aspect/source/simulator/core.cc:2037
#21 0x000055555a183d88 in run_simulator<2> (
    raw_input_as_string="# A test for periodic boundary conditions with particles\n\n# MPI: 2\n\ninclude $ASPECT_SOURCE_DIR/tests/particle_periodic_boundaries.prm\n\n\nsubsection Discretization\n  set Use discontinuous composition di"..., 
    input_as_string="# A test for periodic boundary conditions with particles\n\n# MPI: 2\n\ninclude /home/renegassmoeller/software/aspect/tests/particle_periodic_boundaries.prm\n\n\nsubsection Discretization\n  set Use discontin"..., output_xml=false, output_plugin_graph=false, validate_only=false)
    at /home/renegassmoeller/software/aspect/source/main.cc:609
#22 0x000055555a13f057 in main (argc=2, argv=0x7fffffffcee8) at /home/renegassmoeller/software/aspect/source/main.cc:796
```

**Comment by tjhei** [comment]
Thank you for checking, Rene. The PR is a step forward so we can merge. I am not aware of another issue so I will have to investigate. Traveling this week though.

**Pull request merged**